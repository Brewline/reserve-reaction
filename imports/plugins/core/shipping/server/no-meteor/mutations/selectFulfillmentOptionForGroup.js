import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import getCartById from "../util/getCartById";

const inputSchema = new SimpleSchema({
  cartId: String,
  cartToken: {
    type: String,
    optional: true
  },
  fulfillmentGroupId: String,
  fulfillmentMethodId: String
});

/**
 * @method selectFulfillmentOptionForGroup
 * @summary Selects a fulfillment option for a fulfillment group
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - an object of all mutation arguments that were sent by the client
 * @param {String} input.cartId - The ID of the cart to select a fulfillment option for
 * @param {String} [input.cartToken] - The token for the cart, required if it is an anonymous cart
 * @param {String} input.fulfillmentGroupId - The group to select a fulfillment option for
 * @param {String} input.fulfillmentMethodId - The fulfillment method ID from the option the shopper selected
 * @return {Promise<Object>} An object with a `cart` property containing the updated cart
 */
export default async function selectFulfillmentOptionForGroup(context, input) {
  const cleanedInput = inputSchema.clean(input || {});
  inputSchema.validate(cleanedInput);

  const { cartId, cartToken, fulfillmentGroupId, fulfillmentMethodId } = cleanedInput;
  const { appEvents, collections, userId } = context;
  const { Cart } = collections;

  const cart = await getCartById(context, cartId, { cartToken, throwIfNotFound: true });

  const fulfillmentGroup = (cart.shipping || []).find((group) => group._id === fulfillmentGroupId);
  if (!fulfillmentGroup) throw new ReactionError("not-found", `Fulfillment group with ID ${fulfillmentGroupId} not found in cart with ID ${cartId}`);

  // Make sure there is an option for this group that has the requested ID
  const option = (fulfillmentGroup.shipmentQuotes || []).find((quote) => quote.method._id === fulfillmentMethodId);
  if (!option) throw new ReactionError("not-found", `Fulfillment option with method ID ${fulfillmentMethodId} not found in cart with ID ${cartId}`);

  const { matchedCount } = await Cart.updateOne({
    "_id": cartId,
    "shipping._id": fulfillmentGroupId
  }, {
    $set: {
      "shipping.$.shipmentMethod": option.method
    }
  });
  if (matchedCount !== 1) throw new ReactionError("server-error", "Unable to update cart");

  // We can do the same update locally to avoid a db find
  cart.shipping.forEach((group) => {
    if (group._id === fulfillmentGroupId) {
      group.shipmentMethod = option.method;
    }
  });

  await appEvents.emit("afterCartUpdate", {
    cart,
    updatedBy: userId
  });

  return { cart };
}
