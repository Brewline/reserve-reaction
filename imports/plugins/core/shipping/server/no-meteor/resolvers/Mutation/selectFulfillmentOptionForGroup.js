import { decodeCartOpaqueId, decodeFulfillmentGroupOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/cart";
import { decodeFulfillmentMethodOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/fulfillment";
import selectFulfillmentOptionForGroupMutation from "../../mutations/selectFulfillmentOptionForGroup";

/**
 * @name "Mutation.selectFulfillmentOptionForGroup"
 * @method
 * @memberof Cart/GraphQL
 * @summary resolver for the selectFulfillmentOptionForGroup GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.cartId - The ID of the cart to select a fulfillment option for
 * @param {String} [args.input.cartToken] - The token for the cart, required if it is an anonymous cart
 * @param {String} args.input.fulfillmentGroupId - The group to select a fulfillment option for
 * @param {String} args.input.fulfillmentMethodId - The fulfillment method ID from the option the shopper selected
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} SelectFulfillmentOptionForGroupPayload
 */
export default async function selectFulfillmentOptionForGroup(parentResult, { input }, context) {
  const {
    cartId: opaqueCartId,
    cartToken,
    clientMutationId = null,
    fulfillmentGroupId: opaqueFulfillmentGroupId,
    fulfillmentMethodId: opaqueFulfillmentMethodId
  } = input;

  const cartId = decodeCartOpaqueId(opaqueCartId);
  const fulfillmentGroupId = decodeFulfillmentGroupOpaqueId(opaqueFulfillmentGroupId);
  const fulfillmentMethodId = decodeFulfillmentMethodOpaqueId(opaqueFulfillmentMethodId);

  const { cart } = await selectFulfillmentOptionForGroupMutation(context, {
    cartId,
    cartToken,
    fulfillmentGroupId,
    fulfillmentMethodId
  });

  return {
    cart,
    clientMutationId
  };
}
