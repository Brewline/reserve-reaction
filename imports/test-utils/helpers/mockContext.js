const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
