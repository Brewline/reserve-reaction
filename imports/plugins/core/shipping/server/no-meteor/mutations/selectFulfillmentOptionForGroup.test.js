import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import selectFulfillmentOptionForGroup from "./selectFulfillmentOptionForGroup";

jest.mock("../util/getCartById", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "cartId",
  shipping: [{
    _id: "group1",
    itemIds: ["123"],
    shipmentQuotes: [{
      rate: 0,
      method: {
        _id: "valid-method"
      }
    }],
    type: "shipping"
  }]
})));

const fakeCart = Factory.Cart.makeOne();

test("selects an existing shipping method", async () => {
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(fakeCart));

  const result = await selectFulfillmentOptionForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group1",
    fulfillmentMethodId: "valid-method"
  });
  expect(result).toEqual({
    cart: {
      _id: "cartId",
      shipping: [{
        _id: "group1",
        itemIds: ["123"],
        shipmentQuotes: [{
          rate: 0,
          method: {
            _id: "valid-method"
          }
        }],
        shipmentMethod: {
          _id: "valid-method"
        },
        type: "shipping"
      }]
    }
  });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    "_id": "cartId",
    "shipping._id": "group1"
  }, {
    $set: {
      "shipping.$.shipmentMethod": {
        _id: "valid-method"
      }
    }
  });
});

test("throws if there is no fulfillment group with the given ID", async () => {
  await expect(selectFulfillmentOptionForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group2",
    fulfillmentMethodId: "valid-method"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if there is no fulfillment method with the given ID among the options", async () => {
  await expect(selectFulfillmentOptionForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group1",
    fulfillmentMethodId: "invalid-method"
  })).rejects.toThrowErrorMatchingSnapshot();
});
