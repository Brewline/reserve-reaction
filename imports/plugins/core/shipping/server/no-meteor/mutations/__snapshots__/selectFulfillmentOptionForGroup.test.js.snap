// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if there is no fulfillment group with the given ID 1`] = `"Fulfillment group with ID group2 not found in cart with ID cartId"`;

exports[`throws if there is no fulfillment method with the given ID among the options 1`] = `"Fulfillment option with method ID invalid-method not found in cart with ID cartId"`;
